### ISSUE TITLE
Evalulation is a lot slower than training

### ISSUE THREAD

**Issue by chenchen718** [ppwwyyxx/tensorpack#508]
When training my network, it only takes about 2 hours to complete 1 epoch. But when evaluating, it takes more than 10 hours to process the whole dataset. Does it have something to do with how I read the data? I use multi-process in training and multi-thread in evaluation as suggested. The data reading codes are as follows:

For training
```python
        ds = LMDBData(datadir, shuffle=False)
        ds = PrefetchData(ds, 5000, 1)
        ds = LMDBDataPoint(ds)
        ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
        ds = AugmentImageComponent(ds, augmentors)
        ds = PrefetchDataZMQ(ds, 30)
        ds = BatchData(ds, batch_size)
```

For evaluation
```python
        ds = LMDBData(datadir, shuffle=False)
        ds = PrefetchData(ds, 5000, 1)
        aug = imgaug.AugmentorList(augmentors)

        def mapf(dp):
            fname, cls = dp
            im = cv2.imdecode(fname, cv2.IMREAD_COLOR)
            im = aug.augment(im)
            return im, cls

        ds = LMDBDataPoint(ds, aug)
        ds = ThreadedMapData(ds, 30, mapf, buffer_size=2000, strict=True)
        ds = PrefetchDataZMQ(ds, 1)
        ds = BatchData(ds, batch_size, remainder=True)
```

Thanks very much!

**Comment by ppwwyyxx** [comment]
There isn't a good setting that works for everyone's machine and you'll need to figure it out by testing.  Things can be very different especially with slow hardware. For example, you use 30 threads but I don't even know whether your machine has 30 cores or not, how can I comment anything on it? 

You can run some benchmarks easily to figure out what is causing the slow down.
See [tutorial on performance tuning](http://tensorpack.readthedocs.io/en/latest/tutorial/performance-tuning.html#improve-dataflow).

**Comment by chenchen718** [comment]
Thanks very much for your suggestion. It's very helpful! I compared the performance of the two cases. The multi-process one has a speed of 87.37it/s and the multi-thread one has only 23it/s.

Now the question is if I can use multi-process to inference on the validation set. In you tutorial, you mention that the multi-process data reading is not suitable for the validation set. Is my multi-process code reading a unique and complete copy of the validation set? If not, is there any way to modify the code to make it usable for the validation set?

**Comment by ppwwyyxx** [comment]
The purpose of benchmark is to figure out __what is the bottleneck__, by changing the pipeline. Only knowing how slow the whole pipeline is doesn't help you improve it.

The [documentation](http://tensorpack.readthedocs.io/en/latest/modules/dataflow.html#tensorpack.dataflow.PrefetchDataZMQ) of `PrefetchDataZMQ` is clear that it requires the underlying dataflow to be fully shuffled.
There are other multi-process dataflow such as `MultiProcessMapData`.

**Comment by chenchen718** [comment]
I'd like to make sure if I correctly understand how multi-process functions like `PrefetchDataZMQ` or `MultiProcessMapData` work. 

Assume that I have A, B, C, D functions to process the data. If I don't use these multi-process functions, then all  A, B, C, D would run sequentially in one process, which is obviously slow. If I add a multi-process function to one of the four, let's say A, then A would run in a separate process in parallel with the other 3. That means A doesn't have to wait for B,C,D to finish before starting a new round of work. If I add  a multi-process function to each of them, then all the four can run in parallel. If this is the case, the best way to use multi-process should be adding a multi-process function to any computationally intensive step.

I would be sorry that if this question sounds too basic, but I would like to hear about your explanation before I can step forward with my project on tensorpack. Appreciate your help a lot!

**Comment by ppwwyyxx** [comment]
multiprocessing is just a way a program runs but how it can be used is up to the user. What you described sounds like `MultiProcessMapData` but that's very different from `PrefetchData` as you can see from their APIs.

> Assume that I have A, B, C, D functions to process the data.
I assume you mean "map" the data, i.e. you have x and you need D(C(B(A(x)))). Then yes, you can use multiple processes to run the mapping on a lot of `x`s simultaneously, to perhaps improve speed, if you can beat the overhead of processes. The overhead is mainly on copying the data to processes and copying the result back.

> That means A doesn't have to wait for B,C,D to finish before starting a new round of work. 
Perhaps you mean BCD don't have to wait for A? Not true because logically B always depends on the output of A. But A may finish sooner because you run it multiple processes.

> the best way to use multi-process should be adding a multi-process function to any computationally intensive step.
Not true. Using processes has significant communication overhead so you can't use it everywhere. If you have four mapper ABCD to apply, it's better to merge into one function f(x) = D(C(B(A(x)))) and run f(x) in multiple processes so you don't end up doing too much communication.

`PrefetchDataZMQ` is a completely different mechanism. You can think that it makes several copies of a dataflow and run them in multiple processes. When they are running the results are copied back to main process for use.

**Comment by chenchen718** [comment]
Every time `PrefetchDataZMQ` is used, it creates one or more extra processes for the function before it. And the function following `PrefetchDataZMQ` is the main process. Is it correct?
In the following example, there is one process reading data from the disk and a second one doing data mapping. If I don't use `PrefetchDataZMQ`, there would be only one process sequentially handling all the work. Is it what's happening?
```python
        ds = LMDBData(datadir, shuffle=False)
        ds = PrefetchData(ds, 5000, 1)
        ds = LMDBDataPoint(ds, aug)
        ds = ThreadedMapData(ds, 30, mapf, buffer_size=2000, strict=True)
        ds = PrefetchDataZMQ(ds, 1)
        ds = BatchData(ds, batch_size, remainder=True)
```

**Comment by ppwwyyxx** [comment]
The tutorial http://tensorpack.readthedocs.io/en/latest/tutorial/efficient-dataflow.html contains some explanation of "what's going on" that can help you understand. For the code you posted, there are a total of 3 processes because you use two prefetch with nr_proc=1.
Process A reads LMDB and put data in a queue.
Process B decodes data from the queue, run mapping with 25 threads, put results to ZMQ.
Main process takes data from zmq and batch it.

**Comment by ppwwyyxx** [comment]
There is no recent activity so I'm closing it. In general for any performance problems please do some investigation as in http://tensorpack.readthedocs.io/en/latest/tutorial/performance-tuning.html because I'm unable to run your code with your hardware.
Feel free to reopen if you still have questions.
